featurebyte/query_graph/model/entity_relationship_info.py

```diff
@@ -0,0 +1,208 @@
+                entity_id=entity_id,
+                ancestor_or_descendant_ids=set(),
+                depth=0,
+                max_depth=max_depth,
+            )
+
+        return entity_id_to_ancestor_ids
+
+    @classmethod
+    def get_entity_id_to_descendant_ids(
+        cls,
+        relationships_info: List[EntityRelationshipInfo],
+        max_depth: int = ENTITY_RELATIONSHIP_MAX_DEPTH,
+    ) -> Dict[ObjectId, Set[ObjectId]]:
+        """
+        Get entity id to descendant ids mapping
+
+        Parameters
+        ----------
+        relationships_info: List[EntityRelationshipInfo]
+            Entity relationship info
+        max_depth: int
+            Maximum depth to search
+
+        Returns
+        -------
+        Dict[ObjectId, List[ObjectId]]
+            Entity id to descendant ids mapping
+        """
+        child_to_parent_entity_ids = cls._get_relation_map(
+            relationships_info, parent_to_child=False
+        )
+        entity_id_to_descendant_ids: Dict[ObjectId, Set[ObjectId]] = defaultdict(set)
+        for entity_id in list(child_to_parent_entity_ids):
+            cls._depth_first_search(
+                ancestors_or_descendants_map=entity_id_to_descendant_ids,
+                relation_map=child_to_parent_entity_ids,
+                entity_id=entity_id,
+                ancestor_or_descendant_ids=set(),
+                depth=0,
+                max_depth=max_depth,
+            )
+
+        return entity_id_to_descendant_ids
+
+    @classmethod
+    def create(
+        cls, relationships_info: List[EntityRelationshipInfo]
+    ) -> "EntityAncestorDescendantMapper":
+        """
+        Create a new EntityAncestorDescendantMapper object from the given relationships info
+
+        Parameters
+        ----------
+        relationships_info: List[EntityRelationshipInfo]
+            Entity relationship info
+
+        Returns
+        -------
+        EntityAncestorDescendantMapper
+        """
+        return EntityAncestorDescendantMapper(
+            entity_id_to_ancestor_ids=cls.get_entity_id_to_ancestor_ids(relationships_info),
+            entity_id_to_descendant_ids=cls.get_entity_id_to_descendant_ids(relationships_info),
+        )
+
+    def reduce_entity_ids(self, entity_ids: List[ObjectId]) -> List[ObjectId]:
+        """
+        Reduce entity IDs to only contain the given entity IDs that are not ancestors of any other entity IDs
+
+        Parameters
+        ----------
+        entity_ids: List[ObjectId]
+            List of entity IDs
+
+        Returns
+        -------
+        List[ObjectId]
+        """
+        all_ancestors_ids = set()
+        for entity_id in entity_ids:
+            all_ancestors_ids.update(self.entity_id_to_ancestor_ids[entity_id])
+
+        reduced_entity_ids = set()
+        for entity_id in entity_ids:
+            if entity_id not in all_ancestors_ids:
+                reduced_entity_ids.add(entity_id)
+        return sorted(reduced_entity_ids)
```

Comment:
This class is refactored from existing classes (no new functionalities introduced in this PR).